src/Solvers/preconditioned_conjugate_gradient_solver.jl

```diff
@@ -1,109 +1,128 @@
+                    is used to infer the `architecture`, `grid`, and to create work arrays
+                    that are used internally by the solver.
 
-    ii = interior_parent_indices(location(tf, 1), topology(grid, 1), grid.Nx, grid.Hx)
-    ji = interior_parent_indices(location(tf, 2), topology(grid, 2), grid.Ny, grid.Hy)
-    ki = interior_parent_indices(location(tf, 3), topology(grid, 3), grid.Nz, grid.Hz)
+* `linear_operation`: Function with signature `linear_operation!(p, y, args...)` that calculates
+                     `A*y` and stores the result in `p` for a "candidate solution `y`. `args...`
+                     are optional positional arguments passed from `solve!(x, solver, b, args...)`.
 
-    dotproduct(x, y) = mapreduce((x, y) -> x * y, +, view(x, ii, ji, ki), view(y, ii, ji, ki))
-    norm(x) = sqrt(mapreduce(x -> x * x, +, view(x, ii, ji, ki)))
+* `maximum_iterations`: Maximum number of iterations the solver may perform before exiting.
 
-    Amatrix_function = parameters.Amatrix_function
-    A(x; args...) = (Amatrix_function(a_res, x, arch, grid, bcs; args...); return a_res)
+* `tolerance`: Tolerance for convergence of the algorithm. The algorithm quits when
+               `norm(A*x - b) < tolerance`.
 
-    reference_pressure_solver = nothing
-    if haskey(parameters, :reference_pressure_solver )
-        reference_pressure_solver = parameters.reference_pressure_solver
-    end
+* `precondition`: Function with signature `preconditioner!(z, y, args...)` that calculates
+                    `P*y` and stores the result in `z` for linear operator `P`.
+                    Note that some precondition algorithms describe the step
+                    "solve `M*x = b`" for precondition `M`"; in this context,
+                    `P = M⁻¹`.
 
-    settings = (
-                                q = q,
-                                p = p,
-                                z = z,
-                                r = r,
-                                x = x,
-                              RHS = RHS,
-                              bcs = bcs,
-                             grid = grid,
-                                A = A,
-                                M = M,
-                            maxit = maxit,
-                              tol = tol,
-                          dotprod = dotproduct,
-                             norm = norm,
-                             arch = arch,
-        reference_pressure_solver = reference_pressure_solver,
-    )
-
-   return PreconditionedConjugateGradientSolver(arch, settings)
+See `solve!` for more information about the preconditioned conjugate-gradient algorithm.
+"""
+function PreconditionedConjugateGradientSolver(linear_operation;
+                                               template_field::AbstractField,
+                                               maximum_iterations = prod(size(template_field.grid)),
+                                               tolerance = 1e-13,
```

Comment:
This tolerance (and maximum number of iterations) seems quite restrictive, especially for big problems? I guess you can't pick a default `tolerance` and `maximum_iterations` without some context (and the tolerance is probably quite problem-dependent since it looks like an absolute tolerance).

Maybe the idea is to return a very accurate solution by default?

Reply:
For max iterations: any other number would be arbitrary, wouldn't it? I don't think there's another number we could pick that would always be appropriate. I think it's _conservative_ to set it to something high. `Inf` could be better :-P

As for the `tolerance`, I agree that `1e-13` is arbitrary. We could use something like `eps(eltype(grid))` or `sqrt(eps(eltype(grid))`. I don't think we should use a loose tolerance by default. I think users should set loose tolerances if they need to.

Reply:
Looks like `IterativeSolvers` is a bit looser:

https://julialinearalgebra.github.io/IterativeSolvers.jl/dev/linear_systems/cg/#IterativeSolvers.cg!

they user `maxiter = size(A, 2)` (where `A` is the linear operator) and `tol = sqrt(eps(eltype(grid)))`.

We should also figure out how to use the `IterativeSolvers.jl` implementation :-/ we just have to get broadcasting to work I think (since we already implemented `norm` and `dot`)...

Reply:
Right yeah I'm not sure what the best thing to do here would be, but I agree the current defaults look reasonable.

Seems like the options are:
1. No defaults. Burden of picking a `tolerance` and `max_iterations` falls on the user. Very user unfriendly. But no surprises I guess.
2. Strict tolerance by default. Works out of the box and no burden on users. But if the defaults lead to slow simulations then users will learn this the hard way and will be forced to set looser tolerances. So could also be user unfriendly.
3. Loose tolerance by default. I think this is what MITgcm does? Also works out of the box and no burden on users. Useful default might lead to less user agony as the default might be good for most use cases. But loose default might not be good for all use cases. Seems like the most user friendly, but maybe not for all users?

Reply:
Obviously not something that has to be addressed in this PR by the way.

Reply:
We should turn this into an issue. I don't know what a good default tolerance would be. It's worth discussing. (I think `maximum_iterations` is less important.)

I think if we follow `IterativeSolvers.jl` then with `Float64` we would pick the default `sqrt(eps(Float64)) = 1.49e-8`. I think this is pretty close to what MITgcm uses?